File: js/src/schema/ArgMapper.ts
// Copyright (c) 2018 Spotify AB.
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
import {
  isArray,
  isNumber,
  isBoolean,
  isInteger,
  isUndefined,
  isString
} from '../utils';
import { Command, Node } from '../score';

export default class ArgMapper<T> {
  constructor(
    public readonly name: string,
    public readonly kind: string,
    public readonly defaultValue: T | undefined,
    public readonly kindChecker: (json: any) => boolean
  ) {}

  public static newEnumArg<Enum>(
    name: string,
    defaultValue: Enum,
    possibleValues: string[]
  ): ArgMapper<Enum> {
    const checker = (maybeValue: any) => {
      if (!possibleValues.includes(maybeValue)) {
        throw new Error(
          `arg=${name} expects value=${possibleValues.join(
            '|'
          )} but value is=${maybeValue}`
        );
      }
      return true;
    };
    return new ArgMapper<Enum>(name, 'string', defaultValue, checker);
  }

  public static newStringArg(
    name: string,
    defaultValue?: string
  ): ArgMapper<string> {
    return new ArgMapper<string>(name, 'string', defaultValue, isString);
  }

  public static newIntArg(
    name: string,
    defaultValue?: number
  ): ArgMapper<number> {
    if (!isUndefined(defaultValue) && !isInteger(defaultValue)) {
      throw new Error('expected defaultValue to be an integer');
    }
    return new ArgMapper<number>(name, 'int', defaultValue, isInteger);
  }

  public static newFloatArg(
    name: string,
    defaultValue?: number
  ): ArgMapper<number> {
    return new ArgMapper<number>(name, 'float', defaultValue, isNumber);
  }

  public static newBoolArg(
    name: string,
    defaultValue?: boolean
  ): ArgMapper<boolean> {
    return new ArgMapper<boolean>(name, 'bool', defaultValue, isBoolean);
  }

  public static newTimeArg(
    name: string,
    defaultValue?: number
  ): ArgMapper<number> {
    if (!isUndefined(defaultValue) && !isInteger(defaultValue)) {
      throw new Error('expected defaultValue to be an integer');
    }
    return new ArgMapper<number>(name, 'time', defaultValue, isInteger);
  }

  public static newStringListArg(
    name: string,
    defaultValue?: string[]
  ): ArgMapper<string[]> {
    return this.newListArg(name, 'list(string)', defaultValue, isString);
  }

  public static newIntListArg(
    name: string,
    defaultValue?: number[]
  ): ArgMapper<number[]> {
    if (!isUndefined(defaultValue) && !defaultValue!.every(isInteger)) {
      throw new Error('expected defaultValue to be an integer array');
    }
    return this.newListArg(name, 'list(int)', defaultValue, isInteger);
  }

  public static newFloatListArg(
    name: string,
    defaultValue?: number[]
  ): ArgMapper<number[]> {
    return this.newListArg(name, 'list(float)', defaultValue, isNumber);
  }

  public static newBoolListArg(
    name: string,
    defaultValue?: boolean[]
  ): ArgMapper<boolean[]> {
    return this.newListArg(name, 'list(bool)', defaultValue, isBoolean);
  }

  public static newTimeListArg(
    name: string,
    defaultValue?: number[]
  ): ArgMapper<number[]> {
    if (!isUndefined(defaultValue) && !defaultValue!.every(isInteger)) {
      throw new Error('expected defaultValue to be an integer array');
    }
    return this.newListArg(name, 'list(time)', defaultValue, isInteger);
  }

  private static newListArg<T>(
    name: string,
    kind: string,
    defaultValue: T[] | undefined,
    kindChecker: (json: any) => boolean
  ): ArgMapper<T[]> {
    return new ArgMapper<T[]>(
      name,
      kind,
      defaultValue,
      n => isArray(n) && n.every(kindChecker)
    );
  }

  public getValueOrThrow(value: T | undefined, msg?: string): T {
    if (value === undefined && this.defaultValue !== undefined) {
      return this.defaultValue;
    } else if (value !== undefined) {
      return value;
    } else {
      throw new Error(msg || `expected '${this.name}'`);
    }
  }

  public readConfig(node: Node): T {
    return this.readValue((node.config || {})[this.name]);
  }

  public readArg(command: Command): T {
    return this.readValue((command.args || {})[this.name]);
  }

  private readValue(maybeValue: any | undefined): T {
    if (maybeValue === undefined && this.defaultValue !== undefined) {
      return this.defaultValue;
    } else if (maybeValue === undefined) {
      throw new Error(
        'arg=' + this.name + ' is not set and has no default value'
      );
    }

    if (!this.kindChecker(maybeValue)) {
      throw new Error(
        `arg=${this.name} expects kind=${
          this.kind
        } but value is kind=${typeof maybeValue}`
      );
    }

    return maybeValue as T;
  }
}

File: java/src/main/java/com/spotify/nativeformat/schema/ArgMapper.java
/*-
 * -\-\-
 * nf-grapher-java
 * --
 * Copyright (C) 2016 - 2018 Spotify AB
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * -/-/-
 */

package com.spotify.nativeformat.schema;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.spotify.nativeformat.score.Command;
import com.spotify.nativeformat.score.Node;
import com.spotify.nativeformat.score.Time;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

/**
 * Describes an ArgDef and provides utility methods for reading from and writing to Score objects.
 *
 * @param <T> the real Java type this describes. Should be a type found in Value.
 */
public class ArgMapper<T> {

  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

  private final String name;

  private final String kind;

  private final T defaultValue;

  private final Predicate<JsonNode> kindChecker;

  private final Function<JsonNode, T> valueGetter;

  private ArgMapper(final String name,
                    final String kind,
                    final T defaultValue,
                    final Predicate<JsonNode> kindChecker,
                    final Function<JsonNode, T> valueGetter) {
    this.name = name;
    this.kind = kind;
    this.defaultValue = defaultValue;
    this.kindChecker = kindChecker;
    this.valueGetter = valueGetter;
  }

  public static <T extends Enum> ArgMapper<T> newEnumArg(final String name,
                                                         final T defaultValue,
                                                         final Class<T> enumClass) {
    return new ArgMapper<T>(name, "string", defaultValue, JsonNode::isTextual,
        j -> OBJECT_MAPPER.convertValue(j, enumClass));
  }

  public static ArgMapper<String> newStringArg(final String name, final String defaultValue) {
    return new ArgMapper<>(name, "string", defaultValue, JsonNode::isTextual, JsonNode::asText);
  }

  public static ArgMapper<Long> newIntArg(final String name, final Long defaultValue) {
    return new ArgMapper<>(name, "int", defaultValue, JsonNode::isIntegralNumber, JsonNode::asLong);
  }

  public static ArgMapper<Double> newFloatArg(final String name, final Double defaultValue) {
    return new ArgMapper<>(name, "float", defaultValue, JsonNode::isFloatingPointNumber,
        JsonNode::asDouble);
  }

  public static ArgMapper<Boolean> newBoolArg(final String name, final Boolean defaultValue) {
    return new ArgMapper<>(name, "bool", defaultValue, JsonNode::isBoolean, JsonNode::asBoolean);
  }

  public static ArgMapper<Time> newTimeArg(final String name, final Time defaultValue) {
    return new ArgMapper<>(name, "time", defaultValue, JsonNode::isIntegralNumber,
        (jsonNode) -> OBJECT_MAPPER.convertValue(jsonNode, Time.class));
  }

  public static ArgMapper<List<String>> newStringListArg(final String name,
                                                         final List<String> defaultValue) {
    return newListArg(name, "list(string)", defaultValue,
        JsonNode::isTextual, JsonNode::asText);
  }

  public static ArgMapper<List<Long>> newIntListArg(final String name,
                                                    final List<Long> defaultValue) {
    return newListArg(name, "list(int)", defaultValue,
        JsonNode::isIntegralNumber, JsonNode::asLong);
  }

  public static ArgMapper<List<Double>> newFloatListArg(final String name,
                                                        final List<Double> defaultValue) {
    return newListArg(name, "list(float)", defaultValue,
        JsonNode::isFloatingPointNumber, JsonNode::asDouble);
  }

  public static ArgMapper<List<Boolean>> newBoolListArg(final String name,
                                                        final List<Boolean> defaultValue) {
    return newListArg(name, "list(bool)", defaultValue, JsonNode::isBoolean, JsonNode::asBoolean);
  }

  public static ArgMapper<List<Time>> newTimeListArg(final String name,
                                                     final List<Time> defaultValue) {
    return newListArg(name, "list(time)", defaultValue,
        JsonNode::isIntegralNumber,
        (jsonNode) -> OBJECT_MAPPER.convertValue(jsonNode, Time.class));
  }

  private static <T> ArgMapper<List<T>> newListArg(final String name,
                                                   final String expectedType,
                                                   final List<T> defaultValue,
                                                   final Predicate<JsonNode> typeChecker,
                                                   final Function<JsonNode, T> valueGetter) {
    return new ArgMapper<>(
        name,
        expectedType,
        defaultValue,
        n -> n.isArray() && stream(n).allMatch(typeChecker),
        n -> stream(n).map(valueGetter).collect(Collectors.toList())
    );
  }

  /**
   * Returns the given value if it is non-null or the default value if it exists.
   * Throws an exception if there is no non-null value available.
   *
   * @return the default value
   */
  public T getValueOrThrow(final T val) {
    return this.getValueOrThrow(val, "expected '" + this.name + "'");
  }

  /**
   * Returns the given value if it is non-null or the default value if it exists.
   * Throws an exception if there is no non-null value available.
   *
   * @param message the message to include in the exception
   * @return the default value
   */
  public T getValueOrThrow(final T val, final String message) {
    if (val == null && this.defaultValue != null) {
      return this.defaultValue;
    } else if (val != null) {
      return val;
    } else {
      throw new NullPointerException(message);
    }
  }

  public void addToMap(final T value, final Map<String, Object> map) {
    map.put(this.name, OBJECT_MAPPER.valueToTree(value));
  }

  /**
   * Reads the concrete Java value matching this description from the given Node's config.
   *
   * @param node the node to read from.
   * @return a value
   */
  public T readConfig(final Node node) {
    return readValue(Optional.ofNullable(node.config().get(this.name)));
  }

  /**
   * Reads the concrete Java value matching this description from the given Commands's arguments.
   *
   * @param command the command to read from.
   * @return a value
   */
  public T readArg(final Command command) {
    return readValue(Optional.ofNullable(command.args().get(this.name)));
  }

  @SuppressWarnings("unchecked")
  private T readValue(final Optional<Object> maybeValue) {
    if (!maybeValue.isPresent() && defaultValue == null) {
      throw new RuntimeException("arg=" + name + " is not set and has no default value");
    } else if (!maybeValue.isPresent()) {
      return defaultValue;
    }

    final JsonNode value = OBJECT_MAPPER.valueToTree(maybeValue.get());
    if (!kindChecker.test(value)) {
      throw new RuntimeException(String.format(
          "arg=%s expects kind=%s but value is kind=%s", name, kind, value.getNodeType()
      ));
    }

    return valueGetter.apply(value);
  }

  private static Stream<JsonNode> stream(final JsonNode array) {
    return StreamSupport.stream(array.spliterator(), false);
  }
}
